### ISSUE TITLE
azblob: blob.StartCopyFromURL doesn't work with UTF-8 characters in the source blob

### ISSUE THREAD

**Issue by ncw** [Azure/azure-sdk-for-go#19614]
### Bug Report

If I attempt to call `blob.StartCopyFromURL` from `github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob` on a source blob with UTF-8 characters then the call fails with the error.

```
--------------------------------------------------------------------------------
RESPONSE 400: 400 The metadata specified is invalid. It has characters that are not permitted.
ERROR CODE: InvalidMetadata
--------------------------------------------------------------------------------
﻿<?xml version="1.0" encoding="utf-8"?><Error><Code>InvalidMetadata</Code><Message>The metadata specified is invalid. It has characters that are not permitted.
RequestId:ad893aa4-c01e-0073-6858-fff064000000
Time:2022-11-23T16:28:54.8170886Z</Message></Error>
--------------------------------------------------------------------------------
```

I think this is because the output of `blob.Client().URL()` is not correct. It might be that there is a better way of constructing the source URL for `StartCopyFromURL` - I don't know, and if there is, it should be in the docs!

Here is a reproducer - you will need to change `account` and `containerName` and make sure there are some credentials in the environment

```go
package main

import (
	"bytes"
	"context"
	"fmt"
	"log"

	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/service"
)

const (
	account       = "https://XXXX.blob.core.windows.net/"
	containerName = "XXXX"
	// This is OK
	//blobName = `hello sausage/Hello/ " ' @ < > & +z.txt`
	// This fails
	blobName    = `世界.txt`
	newBlobName = "copytest2.bin"
)

func main() {
	ctx := context.TODO()

	// authenticate with Azure Active Directory
	// this picks up the az login credentials
	cred, err := azidentity.NewDefaultAzureCredential(nil)
	if err != nil {
		log.Fatalf("auth failed: %v", err)
	}

	// create a client for the specified storage account
	client, err := service.NewClient(account, cred, nil)
	if err != nil {
		log.Fatalf("create client failed: %v", err)
	}

	// Make data to upload
	var N int64 = 1024
	buf := make([]byte, N)
	in := bytes.NewReader(buf)

	fmt.Printf("Uploading blob %q, size %d\n", blobName, N)

	cnt := client.NewContainerClient(containerName)
	srcBlob := cnt.NewBlockBlobClient(blobName)

	// upload the file to the specified container with the specified blob name
	_, err = srcBlob.UploadStream(ctx, in, nil)
	if err != nil {
		log.Fatalf("upload failed %v", err)
	}

	// Copy the blob
	fmt.Printf("Calling StartCopyFromURL\n")
	dstBlob := cnt.NewBlockBlobClient(newBlobName)
	srcURL := srcBlob.URL()
	startCopy, err := dstBlob.StartCopyFromURL(ctx, srcURL, nil)
	if err != nil {
		log.Fatalf("start copy failed %v", err)
	}

	fmt.Printf("Copy status is %q\n", *startCopy.CopyStatus)
}
```

This produces

```
Uploading blob "世界.txt", size 1024
Calling StartCopyFromURL
2022/11/23 16:28:54 start copy failed PUT https://XXXX.blob.core.windows.net/XXXX/copytest2.bin
--------------------------------------------------------------------------------
RESPONSE 400: 400 The metadata specified is invalid. It has characters that are not permitted.
ERROR CODE: InvalidMetadata
--------------------------------------------------------------------------------
﻿<?xml version="1.0" encoding="utf-8"?><Error><Code>InvalidMetadata</Code><Message>The metadata specified is invalid. It has characters that are not permitted.
RequestId:ad893aa4-c01e-0073-6858-fff064000000
Time:2022-11-23T16:28:54.8170886Z</Message></Error>
--------------------------------------------------------------------------------
```

Whereas if you remove the UTF-8 characters ``世界` from the source blob name then it works fine.

**PS** this was found by the rclone integration test suite

- import path of package in question: `github.com/Azure/azure-sdk-for-go/sdk/storage/azblob`
- SDK version: This reproduces on the latest release `v0.5.1` and `main` (`v0.5.2-0.20221123052627-63a8c23f3b78`)
- Specify the exact commit if possible: 63a8c23f3b78
- output of `go version`: `go version go1.19.1 linux/amd64`

**Comment by drakkan** [comment]
@ncw I had the same issue, it should work if you manually path escape the blob name, something like this

```
srcBlob := fs.containerClient.NewBlockBlobClient(url.PathEscape(source))
```

This is probably a workaround, no manual escaping was required with the old sdk, but I can live with it

**Comment by ncw** [comment]
Thanks @drakkan - I came up with a very similar workaround, but yours looks slightly neater!

I really think the SDK should be doing this itself though.

**Comment by drakkan** [comment]
> Thanks @drakkan - I came up with a very similar workaround, but yours looks slightly neater!
> 
> I really think the SDK should be doing this itself though.

I think the same workaround applies also to #19613. Basically the sdk join path and query string without escaping them. I think a proper fix should be applied in [JoinPath](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azcore/runtime/request.go#L50)

**Comment by jhendrixMSFT** [comment]
For client constructors that take a blob name, we will now URL path-encode the blobName parameter.  For constructors that take a full blob URL we will leave it as-is (according to the storage folks, this is consistent with how other SDKs behave).

The fix will roll out in `v0.6.0`.
